src/main/scala/sigmastate/eval/RuntimeCosting.scala

```diff
@@ -966,15 +967,26 @@ trait RuntimeCosting extends SigmaLibrary with DataCosting with Slicing { IR: Ev
         }
         withDefaultSize(value, cost)
 
-      case MapCollection(input, id, mapper) =>
+      case MapCollection(input, Terms.Lambda(_, Seq((n, t)), _, Some(mapper))) =>
         val eIn = stypeToElem(input.tpe.elemType)
-        val xs = asRep[CostedCol[Any]](eval(input))
-        implicit val eAny = xs.elem.asInstanceOf[CostedElem[Col[Any],_]].eVal.eA
+        val inputC = evalNode(ctx, env, input).asRep[CostedCol[Any]]
+        implicit val eAny = inputC.elem.asInstanceOf[CostedElem[Col[Any],_]].eVal.eA
+        assert(eIn == eAny, s"Types should be equal: but $eIn != $eAny")
+        val mapperC = fun { x: Rep[Costed[Any]] =>
+          evalNode(ctx, env + (n -> x), mapper)
```

Comment:
@greenhat this `mapperC` should be obtained by recursively calling `eval(f)` where f is second argument.
So only one MapCollection rule is required.

Reply:
@ergomorphic I could not figure out how to do it, however, I did manage to combine both `MapCollection` and `Fold` into single cases handling both `Lambda` and `FuncValue`.